ailiuwSoloProject/ailiuWU_CUI_Project/Market.js

```diff
@@ -0,0 +1,246 @@
+									//console.log(productPrice+productUnit)
+								}
+							}
+							//console.log(jsonContent["entities"][0])
+							cartProduct.push(product)
+							cartProductNum.push(productNum)
+						}
+						else{
+							//add up product number
+							cartProductNum[cartProduct.indexOf(product)]+=productNum
+						}
+						//console.log(product+productNum)
+						//ask for check out or continue shopping
+						return dialog.question({
+									question: 'Would you like to view your cart?',
+									defaultValue: 'no'
+								}).then(viewCart)
+						
+					case 'Get Instructions':
+						//print instructions now
+						instructions()
+						return askForIntention().then(customerIntents)
+					case 'Check Out':
+						return viewCart('yes')
+					case 'View cart':
+						return viewCart('yes')
+					default:
+						return dialog.question(`I don't understand '`+answer+`'. Please re-enter what would you like to do next`)
+									.then(customerIntents)
+				resolve()
+				
+				}
+				
+				
+			})
+						 
+			request.on('error', function(error) {
+				console.log(error)
+				reject()
+			})
+					
+			request.end()
+		})
+}
+
+/*view shopping cart*/
+const viewCart=(answer)=>{
+	if(answer=='yes'){
+		//print cart to the console
+		console.log('Here are the products in your cart:')
+		console.log(sprintf("%-20s%-10s%-10s%-10s", 'Product','Unit','UnitPrice','Number'))
+		for(let i=0;i<cartProduct.length;i++){
+			console.log(sprintf("%-20s%-10s%-10f%-10s",cartProduct[i],cartProductUnit[i],cartProductPrice[i],cartProductNum[i]))
+		}
+		console.log('')
+		return dialog.question({
+			question: 'Would you like to check out?',
+			defaultValue: 'yes'
+		}).then(checkout)
+	}
+	else{
+		console.log('It seems you want to continue shopping')
+		console.log('')
+		//continue shopping
+		return askForIntention().then(customerIntents)
+	}
+}
+
+/*check out*/
+const checkout=(answer)=>{
+	if(answer=='yes'){
+		console.log('Here is your receipt:')
+		console.log(sprintf("%-20s%-10s%-10s%-10s%-10s", 'Product','Unit','UnitPrice','Number','SubTotal'))
+		for(let i=0;i<cartProduct.length;i++){
+			let subTotal=cartProductPrice[i]*cartProductNum[i]
+			console.log(sprintf("%-20s%-10s%-10f%-10s%-10f",cartProduct[i],cartProductUnit[i],cartProductPrice[i],cartProductNum[i],subTotal))
+			totalCost+=subTotal
+		}
+		console.log('Your total cost for today is :'+totalCost)			
+		console.log('')
+		console.log('Thank you for shopping at our market!')
+		console.log('Have a nice day!')
+		//complete this adventure
+		dialog.complete()
+	}
+	else{
+		console.log('It seems you want to continue shopping')
+		console.log('')
+		//continue shopping
+		return askForIntention().then(customerIntents)
+	}
+}
+
+
+/*Main program*/
+console.log('Hello! Welcome to the market!')
+
+greetings()
+  .then(askForIntention)
+  .then(customerIntents)
+  .catch((err) => {
+	  //print error	  
+      console.log(err)
+	  //complete this adventure
+      dialog.complete()
+    })
+	
+}
```

Comment:
Given that you fit all of this into ~250 lines of code, I think there is a reasonable argument to be made that it makes sense not to follow the Single Responsibility Principle. However, if we wanted to follow it, here is one way you could segregate this code into smaller files that meet SRP:

* Market
  * facilitates the terminal dialog
* addToCart
  * add to cart intent handling
* viewCart
  * view cart intent handling
* help
  * help intent handling
* checkout
  * checkout intent handling
* dialogflowClient
  * integration with DialogFlow (encapsulates the details, and just returns a simple interface like, `parseIntent`)
* cartRepository
  * this would encapsulate the arrays you used to store the cart, and could be refactored to use another type of storage, such as a database, without having to refactor other code